ambari-server/src/main/java/org/apache/ambari/server/api/services/ClusterService.java

```diff
@@ -860,4 +879,81 @@ ResourceInstance createArtifactResource(String clusterName, String artifactName)
 
     return createResource(Resource.Type.Artifact, mapIds);
   }
+
+  private boolean parseBody(String body) {
+    JsonObject jsonObject = gson.fromJson(body, JsonObject.class);
+    if (jsonObject == null) {
+      return false;
+    }
+
+    JsonElement artifact_data = jsonObject.get("artifact_data");
+    if (artifact_data == null) {
+      return false;
+    }
+    JsonArray identities = artifact_data.getAsJsonObject().getAsJsonArray("identities");
+    if (identities != null) {
+      if (checkKeytabsPrincipal(identities)) {
+        return true;
+      }
+    }
+
+    JsonArray services = artifact_data.getAsJsonObject().getAsJsonArray("services");
+    for (int i = 0; i < services.size(); i++) {
+      JsonArray components = services.get(i).getAsJsonObject().get("components").getAsJsonArray();
```

Comment:
We can simplify all get(memberName).getAsJsonArray() to getAsJsonArray(memberName) and  checks for the existence of JSON fields before accessing them

Reply:
Object level mapping is required to to check/validate the principal and keytab values. Also code readability is more in this way.